Tweak comma handling of "missing match arm" suggestion and fix "remove this arm" suggestion, and make suggestion verbose

estebank opened the pull request:
Better track trailing commas in match arms. Do not suggest adding trailing comma to match arm with block body. Better heuristic for "is this match in one line".

When encountering an unreachable match arm, (correctly) suggest removing the entire arm:

```
error: a never pattern is always unreachable
  --> $DIR/ICE-130779-never-arm-no-oatherwise-block.rs:10:20
   |
LL |         ! if true => {}
   |                      ^^ this will never be executed
   |
help: remove the unreachable match arm
   |
LL -         ! if true => {}
   |
```

Noticed in https://github.com/rust-lang/rust/pull/137343#discussion_r1964304010.

r? @compiler-errors

The first commit is independent of the second, but to make the second one produce accurate suggestions the span needs to include the trailing comma, hence the grouping of both changes in this PR.

estebank committed:
Tweak "add match arm" suggestion

Better track trailing commas in match arms. Do not suggest adding trailing comma to match arm with block body. Better heuristic for "is this match in one line".

estebank committed:
Fix incorrect suggestion for never patterns with bodies or guard

When encountering an unreachable match arm, (correctly) suggest removing the entire arm:

```
error: a never pattern is always unreachable
  --> $DIR/ICE-130779-never-arm-no-oatherwise-block.rs:10:20
   |
LL |         Some(!) if true => {}
   |                            ^^ this will never be executed
   |
help: remove the unreachable match arm
   |
LL -         Some(!) if true => {}
LL +         Some(!),
   |
```

Noticed in https://github.com/rust-lang/rust/pull/137343#discussion_r1964304010.

estebank commented on tests/ui/pattern/usefulness/empty-match.exhaustive_patterns.stderr:
```diff
@@ -176,9 +176,8 @@ LL |     match_guarded_arm!(0u8);
-LL ~                 _ if false => {},
-LL +                 0_u8..=u8::MAX => todo!()
-   |
+LL |                 _ if false => {}, 0_u8..=u8::MAX => todo!()
+   |                                 +++++++++++++++++++++++++++
```
This suggestion is *inside* a macro definition. Arguably, we shouldn't emit a suggestion for that, specially if the macro is non-local, but having a span pointing at the relevant `match` would still be useful for macro writers. For now I didn't touch that logic (beyond the different span contexts making the suggestion assume it is in a single line instead of multiple).

estebank replied:
If anything, the only reasonable suggestion we could provide within a macro is `_ => panic!()`.

estebank replied:
We now detect the macro case and don't provide a suggestion, only show appropriate spans.

estebank commented on tests/ui/pattern/usefulness/empty-types.never_pats.stderr:
```diff
@@ -430,7 +430,7 @@ note: `Option<!>` defined here
 help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
    |
-LL ~         &None => {},
+LL ~         &None => {}
 LL +         &Some(!)
```
Another case of already broken suggestion :dizzy_face:

estebank replied:
Given this, the second commit is incorrect. Fixing.

estebank replied:
Never types sure need some polish :-/

Except for better handling cases involving macros in patterns and or-patterns surrounded by parentheses (which is recovered syntax to begin with), the suggestions are now "correct".

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13468222659/job/37638077621) [(plain)](https://github.com/rust-lang/rust/commit/d112fca1326e06767d27a7b365f5bf6e60cdfa11/checks/37638077621/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/ui/ref_option/ref_option.all.fixed ... ok
tests/ui/non_std_lazy_static/non_std_lazy_static_fixable.fixed ... ok

FAILED TEST: tests/ui/match_same_arms_non_exhaustive.rs
command: CLIPPY_CONF_DIR="tests" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-e6ea37cb638e63c2.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-d76aca92153d9a0d.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-2fc2dfabb8d9f9a4.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-c6b15f405131634a.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-45ef8a96f3e545c7.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-9656da8cc10b9684.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-f91003c2878d7b70.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-fcde03586ffe7a98.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-96712d1447082581.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-dab342f989862ee6.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserde_derive-d1c9d4326af97dc8.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-0bb941cb3a5dd37e.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-a8ed6d19d4a7e1b0.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--crate-type=lib" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui" "tests/ui/match_same_arms_non_exhaustive.rs" "--edition" "2021"
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/match_same_arms_non_exhaustive.stderr` to the actual output
--- tests/ui/match_same_arms_non_exhaustive.stderr
+++ <stderr output>
+++ <stderr output>
 error: this match arm has an identical body to the `_` wildcard arm
... 8 lines skipped ...
    |
 LL |         _ => repeat(),
-   |         ^^^^^^^^^^^^^
---



FAILED TEST: tests/ui/match_same_arms.rs
command: CLIPPY_CONF_DIR="tests" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-e6ea37cb638e63c2.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-d76aca92153d9a0d.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-2fc2dfabb8d9f9a4.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-c6b15f405131634a.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-45ef8a96f3e545c7.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-9656da8cc10b9684.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-f91003c2878d7b70.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-fcde03586ffe7a98.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-96712d1447082581.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-dab342f989862ee6.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserde_derive-d1c9d4326af97dc8.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-0bb941cb3a5dd37e.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-a8ed6d19d4a7e1b0.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui" "tests/ui/match_same_arms.rs" "--edition" "2021"
error: actual output differed from expected
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/match_same_arms.stderr` to the actual output
--- tests/ui/match_same_arms.stderr
+++ <stderr output>
 error: this match arm has an identical body to the `_` wildcard arm
... 7 lines skipped ...
    |
 LL |         _ => 0,
-   |         ^^^^^^
-   |         ^^^^^^
+   |         ^^^^^^^
    = note: `-D clippy::match-same-arms` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::match_same_arms)]`
... 3 lines skipped ...
    |
 LL |         (1, .., 3) => 42,
+   |         ^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 8 lines skipped ...
---



FAILED TEST: tests/ui/match_same_arms2.rs
command: CLIPPY_CONF_DIR="tests" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-e6ea37cb638e63c2.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-d76aca92153d9a0d.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-2fc2dfabb8d9f9a4.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-c6b15f405131634a.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-45ef8a96f3e545c7.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-9656da8cc10b9684.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-f91003c2878d7b70.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-fcde03586ffe7a98.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-96712d1447082581.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-dab342f989862ee6.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserde_derive-d1c9d4326af97dc8.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-0bb941cb3a5dd37e.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-a8ed6d19d4a7e1b0.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui" "tests/ui/match_same_arms2.rs" "--edition" "2021"
error: actual output differed from expected
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/match_same_arms2.stderr` to the actual output
--- tests/ui/match_same_arms2.stderr
+++ <stderr output>
 error: this match arm has an identical body to the `_` wildcard arm
... 19 lines skipped ...
 LL | |             a
 LL | |         },
-   | |_________^
---
    |
    = help: try changing either arm body
... 9 lines skipped ...
    |
 LL |         (None, Some(a)) => bar(a),
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 9 lines skipped ...
... 9 lines skipped ...
    |
 LL |         (None, Some(a)) if a == 42 => a,
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 9 lines skipped ...
... 9 lines skipped ...
    |
 LL |         (Some(a), ..) => bar(a),
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 8 lines skipped ...
... 8 lines skipped ...
    |
 LL |         (Ok(x), Some(_)) => println!("ok {}", x),
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 8 lines skipped ...
... 8 lines skipped ...
    |
 LL |         Ok(_) => println!("ok"),
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 11 lines skipped ...
---
    |
    = help: try changing either arm body
... 11 lines skipped ...
    |
 LL |         Foo::X(0) => 1,
+   |         ^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 9 lines skipped ...
... 9 lines skipped ...
    |
 LL |         Foo::Z(_) => 1,
+   |         ^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 8 lines skipped ...
... 8 lines skipped ...
    |
 LL |         Some(Bar { y: 0, x: 5, .. }) => 1,
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 9 lines skipped ...
... 9 lines skipped ...
    |
 LL |         1 => cfg!(not_enable),
+   |         ^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 9 lines skipped ...
... 9 lines skipped ...
    |
 LL |                 MaybeStaticStr::Borrowed(s) => s,
+   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 7 lines skipped ...
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13468734798/job/37639354836) [(plain)](https://github.com/rust-lang/rust/commit/15b16fe4ad7e0222b296058bfa757630bb7d1457/checks/37639354836/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/ui/ref_option/ref_option.all.fixed ... ok
tests/ui/non_std_lazy_static/non_std_lazy_static_fixable.fixed ... ok

FAILED TEST: tests/ui/match_same_arms_non_exhaustive.rs
command: CLIPPY_CONF_DIR="tests" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-e6ea37cb638e63c2.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-d76aca92153d9a0d.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-2fc2dfabb8d9f9a4.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-c6b15f405131634a.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-45ef8a96f3e545c7.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-9656da8cc10b9684.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-f91003c2878d7b70.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-fcde03586ffe7a98.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-96712d1447082581.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-dab342f989862ee6.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserde_derive-d1c9d4326af97dc8.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-0bb941cb3a5dd37e.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-a8ed6d19d4a7e1b0.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--crate-type=lib" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui" "tests/ui/match_same_arms_non_exhaustive.rs" "--edition" "2021"
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/match_same_arms_non_exhaustive.stderr` to the actual output
--- tests/ui/match_same_arms_non_exhaustive.stderr
+++ <stderr output>
+++ <stderr output>
 error: this match arm has an identical body to the `_` wildcard arm
... 8 lines skipped ...
    |
 LL |         _ => repeat(),
-   |         ^^^^^^^^^^^^^
---



FAILED TEST: tests/ui/match_same_arms.rs
command: CLIPPY_CONF_DIR="tests" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-e6ea37cb638e63c2.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-d76aca92153d9a0d.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-2fc2dfabb8d9f9a4.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-c6b15f405131634a.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-45ef8a96f3e545c7.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-9656da8cc10b9684.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-f91003c2878d7b70.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-fcde03586ffe7a98.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-96712d1447082581.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-dab342f989862ee6.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserde_derive-d1c9d4326af97dc8.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-0bb941cb3a5dd37e.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-a8ed6d19d4a7e1b0.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui" "tests/ui/match_same_arms.rs" "--edition" "2021"
error: actual output differed from expected
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/match_same_arms.stderr` to the actual output
--- tests/ui/match_same_arms.stderr
+++ <stderr output>
 error: this match arm has an identical body to the `_` wildcard arm
... 7 lines skipped ...
    |
 LL |         _ => 0,
-   |         ^^^^^^
-   |         ^^^^^^
+   |         ^^^^^^^
    = note: `-D clippy::match-same-arms` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::match_same_arms)]`
... 3 lines skipped ...
    |
 LL |         (1, .., 3) => 42,
+   |         ^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 8 lines skipped ...
---



FAILED TEST: tests/ui/match_same_arms2.rs
command: CLIPPY_CONF_DIR="tests" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-e6ea37cb638e63c2.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-d76aca92153d9a0d.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-2fc2dfabb8d9f9a4.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-c6b15f405131634a.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-45ef8a96f3e545c7.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-9656da8cc10b9684.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-f91003c2878d7b70.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-fcde03586ffe7a98.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-96712d1447082581.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-dab342f989862ee6.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserde_derive-d1c9d4326af97dc8.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-0bb941cb3a5dd37e.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-a8ed6d19d4a7e1b0.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui" "tests/ui/match_same_arms2.rs" "--edition" "2021"
error: actual output differed from expected
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/match_same_arms2.stderr` to the actual output
--- tests/ui/match_same_arms2.stderr
+++ <stderr output>
 error: this match arm has an identical body to the `_` wildcard arm
... 19 lines skipped ...
 LL | |             a
 LL | |         },
-   | |_________^
---
    |
    = help: try changing either arm body
... 9 lines skipped ...
    |
 LL |         (None, Some(a)) => bar(a),
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 9 lines skipped ...
... 9 lines skipped ...
    |
 LL |         (None, Some(a)) if a == 42 => a,
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 9 lines skipped ...
... 9 lines skipped ...
    |
 LL |         (Some(a), ..) => bar(a),
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 8 lines skipped ...
... 8 lines skipped ...
    |
 LL |         (Ok(x), Some(_)) => println!("ok {}", x),
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 8 lines skipped ...
... 8 lines skipped ...
    |
 LL |         Ok(_) => println!("ok"),
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 11 lines skipped ...
---
    |
    = help: try changing either arm body
... 11 lines skipped ...
    |
 LL |         Foo::X(0) => 1,
+   |         ^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 9 lines skipped ...
... 9 lines skipped ...
    |
 LL |         Foo::Z(_) => 1,
+   |         ^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 8 lines skipped ...
... 8 lines skipped ...
    |
 LL |         Some(Bar { y: 0, x: 5, .. }) => 1,
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 9 lines skipped ...
... 9 lines skipped ...
    |
 LL |         1 => cfg!(not_enable),
+   |         ^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 9 lines skipped ...
... 9 lines skipped ...
    |
 LL |                 MaybeStaticStr::Borrowed(s) => s,
+   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: try changing either arm body
... 7 lines skipped ...
```

</details>

estebank committed:
fix clippy tests

estebank committed:
Account for non-exhausitve match expression coming from macro expansion

```
error[E0004]: non-exhaustive patterns: `u8::MAX` not covered
  --> $DIR/exhaustiveness.rs:47:8
   |
LL |     m!(0u8, 0..255);
   |        ^^^ pattern `u8::MAX` not covered
   |
   = note: the matched value is of type `u8`
note: within macro `m`, this `match` expression doesn't expand to cover all patterns
  --> $DIR/exhaustiveness.rs:7:9
   |
LL | / macro_rules! m {
LL | |     ($s:expr, $($t:tt)+) => {
LL | |         match $s { $($t)+ => {} }
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^^
LL | |     }
LL | | }
   | |_-
   = help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern
```

estebank committed:
When lowering paren surrounded pattern, keep outer span

Parentheses surrounding patterns are not ketp in the HIR (and are warned against in the AST). In order to avoid having some suggestions break, we keep the outer span (including parentheses) when lowering the patterns.

```
error: a never pattern is always unreachable
  --> $DIR/ICE-133117-duplicate-never-arm.rs:9:26
   |
LL |         (!|!) if true => {}
   |                          ^^ this will never be executed
   |
help: remove the match arm expression
   |
LL -         (!|!) if true => {}
LL +         (!|!),
   |
```
```
error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:135:15
   |
LL |     let [&mut &(mut x)] = &mut [&0];
   |               ^^^^^^^^    --------- this expression has type `&mut [&{integer}; 1]`
   |               |
   |               expected integer, found `&_`
   |
   = note:   expected type `{integer}`
           found reference `&_`
help: consider removing `&` from the pattern
   |
LL -     let [&mut &(mut x)] = &mut [&0];
LL +     let [&mut (mut x)] = &mut [&0];
   |
```

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13532718128/job/37818364331) [(plain)](https://github.com/rust-lang/rust/commit/2b9811685f6bffba63776474e71af3cc2e3b89d1/checks/37818364331/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/ui/ref_option/ref_option.all.fixed ... ok
tests/ui/non_std_lazy_static/non_std_lazy_static_fixable.fixed ... ok

FAILED TEST: tests/ui/single_element_loop.rs
command: CLIPPY_CONF_DIR="tests" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-e6ea37cb638e63c2.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-d76aca92153d9a0d.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-2fc2dfabb8d9f9a4.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-c6b15f405131634a.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-45ef8a96f3e545c7.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-9656da8cc10b9684.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-f91003c2878d7b70.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libquote-fcde03586ffe7a98.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libregex-96712d1447082581.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libserde-dab342f989862ee6.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserde_derive-d1c9d4326af97dc8.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-0bb941cb3a5dd37e.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-a8ed6d19d4a7e1b0.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/release/deps" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/ui_test/tests/ui" "tests/ui/single_element_loop.rs" "--edition" "2021"
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/single_element_loop.stderr` to the actual output
--- tests/ui/single_element_loop.stderr
+++ <stderr output>
+++ <stderr output>
 error: for loop over a single element
   --> tests/ui/single_element_loop.rs:8:5
... 92 lines skipped ...
    |
 LL ~     {
-LL +         let (Ok(mut _x) | Err(mut _x)) = res_void;
+LL +         let ((Ok(mut _x) | Err(mut _x))) = res_void;
 LL +         let ptr: *const bool = std::ptr::null();
 LL +     }
 error: aborting due to 8 previous errors
 


```

</details>

estebank committed:
Fallback removal span when involving macros

```
error: a never pattern is always unreachable
  --> $DIR/pattern-behind-macro.rs:13:21
   |
LL |         never!() => {}
   |                     ^^ this will never be executed
   |
help: remove the match arm expression
   |
LL -         never!() => {}
LL +         never!(),
   |
```

Look up the macro backtrace call sites to see if we find where the macro was used as a pattern, to properly suggest removing match arm guard and body.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13534528238/job/37823625476) [(plain)](https://github.com/rust-lang/rust/commit/c35c3bb8bf394cc2bc2a717174ef43842b70c6e8/checks/37823625476/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/ui/ref_option/ref_option.all.fixed ... ok
tests/ui/non_std_lazy_static/non_std_lazy_static_fixable.fixed ... ok

FAILED TEST: tests/ui/single_element_loop.rs
command: CLIPPY_CONF_DIR="tests" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-e6ea37cb638e63c2.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-d76aca92153d9a0d.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-2fc2dfabb8d9f9a4.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-c6b15f405131634a.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-45ef8a96f3e545c7.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-9656da8cc10b9684.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-f91003c2878d7b70.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libquote-fcde03586ffe7a98.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libregex-96712d1447082581.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libserde-dab342f989862ee6.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserde_derive-d1c9d4326af97dc8.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-0bb941cb3a5dd37e.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-a8ed6d19d4a7e1b0.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/release/deps" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/ui_test/tests/ui" "tests/ui/single_element_loop.rs" "--edition" "2021"
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/single_element_loop.stderr` to the actual output
--- tests/ui/single_element_loop.stderr
+++ <stderr output>
+++ <stderr output>
 error: for loop over a single element
   --> tests/ui/single_element_loop.rs:8:5
... 92 lines skipped ...
    |
 LL ~     {
-LL +         let (Ok(mut _x) | Err(mut _x)) = res_void;
+LL +         let ((Ok(mut _x) | Err(mut _x))) = res_void;
 LL +         let ptr: *const bool = std::ptr::null();
 LL +     }
 error: aborting due to 8 previous errors
 


```

</details>

compiler-errors commented on compiler/rustc_ast_lowering/src/expr.rs:
```diff
@@ -676,6 +676,23 @@ impl<'hir> LoweringContext<'_, 'hir> {
         {
             self.lower_expr(body)
         } else {
+            let removal_span = |mut removal_span: Span| {
```
Is this necessary? Can you use `find_ancestor_in_same_ctxt` instead of cooking it yourself?

compiler-errors replied:
You should be able to do like... `pat_span.find_ancestor_in_same_ctxt(arm_span)`.

estebank replied:
Done (amended last commit).

compiler-errors commented:
r=me when CI is green

compiler-errors commented:
Also @estebank please for this PR and all others in the future, can you make sure to note that you're making these suggestions verbose. Seems like a major thing to leave out when that's presumably your underlying motivation here. Thx

estebank commented on tests/ui/pattern/bindings-after-at/borrowck-pat-ref-mut-and-ref.stderr:
```diff
@@ -10,18 +10,18 @@ error: cannot borrow value as mutable more than once at a time
 LL |     let ref mut a @ (ref b @ ref mut c) = u(); // sub-in-sub
-   |         ^^^^^^^^^    -----   --------- value is mutably borrowed by `c` here
-   |         |            |
-   |         |            value is borrowed by `b` here
+   |         ^^^^^^^^^   ------   --------- value is mutably borrowed by `c` here
+   |         |           |
+   |         |           value is borrowed by `b` here
```
This is technically a regression that I haven't figured out how to get rid of without breaking anything else :-/

bors commented:
:pushpin: Commit 88398eb28d35db329ad2ba69d8bd40b76c83e376 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 88398eb28d35db329ad2ba69d8bd40b76c83e376 -->
<!-- homu: {"type":"Approved","sha":"88398eb28d35db329ad2ba69d8bd40b76c83e376","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r-
needs rebase for tests/ui/never_type/unused_trait_in_never_pattern_body.rs from what I see

estebank committed:
fix rebase

compiler-errors commented on tests/ui/never_type/unused_trait_in_never_pattern_body.stderr:
```diff
@@ -18,10 +18,18 @@ LL | |
+   |
+help: remove the match arm expression
+   |
+LL -         ! => || {
+LL -
+LL -
+LL -             use std::ops::Add;
+LL -             0.add(1)
+LL -         },
+LL +         !,
```
This is a lot of lines, to the point that it's a bit hard to even see (without coloration) that this is even suggesting to leave the plain `!` and remove everything else. we should probably make this suggestion hidden?

estebank committed:
Make never pattern suggestions hidden

estebank commented on tests/ui/rfcs/rfc-0000-never_patterns/ICE-130779-never-arm-no-oatherwise-block.stderr:
```diff
@@ -16,10 +16,9 @@ error: a never pattern is always unreachable
-   |                    help: remove this expression
+   |                    ^^ this will never be executed
+   |
+   = help: remove the match arm expression
```
This probably  be  "remove the match arm guard and expression" now.

estebank commented on tests/ui/rfcs/rfc-0000-never_patterns/ICE-130779-never-arm-no-oatherwise-block-fix.stderr:
```diff
@@ -0,0 +1,24 @@
+   |         |
+   |         while parsing this or-pattern starting here
+   |
+help: remove the `|`
+   |
+LL -         ! |
+LL +         !
+   |
```
Fix for the left-over whitespace coming soon

compiler-errors commented:
Do you plan on tweaking this further? If not, please open a tracking issue (https://rust-lang.zulipchat.com/#narrow/channel/147480-t-compiler.2Fwg-diagnostics/topic/Why.20are.20we.20making.20suggestions.20verbose.3F/near/502691049) and it would be nice to cc that on this PR and any further PRs having to do with verbose-ifying diagnostics. r=me afterwards.

estebank commented:
@bors r=compiler-errors

bors commented:
:pushpin: Commit 7e1f8f2106e8389fb1f6d87b101bd07c97dd1e0f has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 7e1f8f2106e8389fb1f6d87b101bd07c97dd1e0f -->
<!-- homu: {"type":"Approved","sha":"7e1f8f2106e8389fb1f6d87b101bd07c97dd1e0f","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/137373#issuecomment-2694703372) for pull requests below priority 102. This pull request will be tested once the tree is reopened.

compiler-errors commented:
@bors r-

needs tracking issue
